iarc_forebrain/scripts/Drone.py

```diff
@@ -0,0 +1,518 @@
+import math
+
+import rospy
+import tf
+import tf.transformations
+from ardrone_autonomy.msg import Navdata
+from geometry_msgs.msg import Twist, PoseStamped, Pose, Point, Quaternion
+from iarc_arbiter.msg import RegisterBehavior, VelAlt
+from iarc_main.msg import Roomba
+from numpy import pi
+from std_msgs.msg import String, Empty, Header
+
+
+class Drone:
+    ROOMBA_HEIGHT = 0.1  # Height of a Roomba top in meters
+    MAX_VERTICAL_VEL = 2.0
+    MIN_VERTICAL_VEL = 0.8
+    NORMAL_HEIGHT = 2.5
+
+    def __init__(self, target=None, tfl=None):
+        """
+        :type tfl: tf.TransformListener
+        """
+        self.should_hit_button = False
+        self.should_land_front = False
+        self.vel3d = Twist()
+
+        # World state
+        self._remembers_flying = False
+
+        # Remembered control state
+        self.last_height = 0.0
+        self.current_target = target  # Roomba class
+
+        self.prev_target_facing_angle = None  # in radian
+
+        self.FRAME_ID = "base_link"
+
+        if tfl is None:
+            self.tf = tf.TransformListener()
+        else:
+            self.tf = tfl
+
+        self.posPub = rospy.Publisher('/forebrain/cmd_pos', PoseStamped, queue_size=0)
+        self.velPub = rospy.Publisher('/forebrain/cmd_vel', Twist, queue_size=0)
+        self.takeoffPub = rospy.Publisher('/forebrain/cmd_takeoff', Empty, queue_size=0)
+        self.landPub = rospy.Publisher('/forebrain/cmd_land', Empty, queue_size=0)
+        self.velAltPub = rospy.Publisher('/forebrain/cmd_vel_alt', VelAlt, queue_size=0)
+
+        rospy.Publisher('/arbiter/register', RegisterBehavior, latch=True, queue_size=10).publish(
+            name='forebrain', fast=True)
+        rospy.sleep(0.1)
+        rospy.Publisher('/arbiter/activate_behavior', String, latch=True, queue_size=10).publish(
+            'forebrain')
+
+        rospy.Subscriber("/cmd_vel", Twist, self.record_vel)
+
+        self.navdata = None
+        rospy.Subscriber('/ardrone/navdata', Navdata, self.onNavdata)
```

Comment:
This seems specific to the Parrot AR Drone. Consider parametrizing or adding a layer of abstraction.